pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -109,17 +119,54 @@ func (r *Reconciler) ValidateClusterSpec(ctx context.Context, log logr.Logger, t
 	return controller.Result{}, nil
 }
 
-// ReconcileControlPlane applies the control plane CAPI objects to the cluster.
-func (r *Reconciler) ReconcileControlPlane(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
+// DetectOperationAndGenerateSpec generates control plane and workers spec, then detect create or upgrade type.
+func (r *Reconciler) DetectOperationAndGenerateSpec(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
 	spec := tinkerbellScope.ClusterSpec
-	log = log.WithValues("phase", "reconcileControlPlane")
-	log.Info("Applying control plane CAPI objects")
+	log = log.WithValues("phase", "DetectOperation")
+	log.Info("Detecting operation type")
+
 	cp, err := tinkerbell.ControlPlaneSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
 	if err != nil {
 		return controller.Result{}, err
 	}
+	tinkerbellScope.controlPlane = cp
+
+	w, err := tinkerbell.WorkersSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
+	if err != nil {
+		return controller.Result{}, errors.Wrap(err, "generating workers spec")
+	}
+	tinkerbellScope.workers = w
+
+	err = detectOperation(ctx, log, tinkerbellScope)
+	if err != nil && apierrors.IsNotFound(err) {
+		return controller.Result{}, errors.Wrap(err, "Detecting operation type")
+	}
+
+	// omit machine template on scale update
+	if tinkerbellScope.operationType == scaleUpdate {
+		tinkerbellScope.controlPlane.ControlPlaneMachineTemplate = nil
+		wg := tinkerbellScope.workers.Groups
```

Comment:
imo this should be part of the tinkerbell package, in the spec generation code

Reply:
sure. I moved it to tinkerbell package but don't let provider consume it yet. I think this could be done in a separate pr, maybe when we clean up all providers and need to have all providers using api cp and workers.

Reply:
[Clean up] All providers use CAPI api#1061](https://github.com/aws/eks-anywhere-internal/issues/1061)